src/rgw/rgw_rest.cc

```diff
@@ -369,63 +370,88 @@ void dump_errno(struct req_state *s, int http_ret)
+                 const long long val)
+{
+  char buf[32];
+  const auto len = snprintf(buf, sizeof(buf), "%lld", val);
+
+  return dump_header(s, name, boost::string_ref(buf, len));
+}
+
+void dump_header(struct req_state* const s,
+                 const boost::string_ref& name,
+                 const utime_t& ut)
+{
+  char buf[32];
+  const auto len = snprintf(buf, sizeof(buf), "%lld.%05d",
+	                    static_cast<long long>(ut.sec()),
+                            static_cast<int>(ut.usec() / 10));
+
+  return dump_header(s, name, boost::string_ref(buf, len));
+}
+
+void dump_content_length(struct req_state* const s, const uint64_t len)
+{
+  try {
+    RESTFUL_IO(s)->send_content_length(len);
+  } catch (RGWRestfulIOEngine::Exception& e) {
     ldout(s->cct, 0) << "ERROR: s->cio->send_content_length() returned err="
-                     << r << dendl;
-  }
-  r = STREAM_IO(s)->print("Accept-Ranges: bytes\r\n");
-  if (r < 0) {
-    ldout(s->cct, 0) << "ERROR: s->cio->print() returned err=" << r << dendl;
+                     << e.value() << dendl;
   }
+  dump_header(s, "Accept-Ranges", "bytes");
 }
 
-void dump_etag(struct req_state * const s, const char * const etag)
+void dump_etag(struct req_state* const s, const boost::string_ref& etag)
 {
-  if ('\0' == *etag) {
+  if (etag.empty()) {
     return;
   }
 
-  int r;
   if (s->prot_flags & RGW_REST_SWIFT) {
-    r = STREAM_IO(s)->print("etag: %s\r\n", etag);
+    return dump_header(s, "etag", etag);
   } else {
-    r = STREAM_IO(s)->print("ETag: \"%s\"\r\n", etag);
-  }
-  if (r < 0) {
-    ldout(s->cct, 0) << "ERROR: s->cio->print() returned err=" << r << dendl;
+    /* We need two extra bytes for quotes. */
+    char buf[etag.size() + 2 + 1];
+    const auto len = snprintf(buf, sizeof(buf), "\"%s\"", etag.data());
+
+    return dump_header(s, "ETag: \"%s\"\r\n", boost::string_ref(buf, len));
   }
 }
 
-void dump_pair(struct req_state *s, const char *key, const char *value)
+void dump_etag(struct req_state* const s, ceph::buffer::list& bl_etag)
 {
-  if ( (strlen(key) > 0) && (strlen(value) > 0))
-    STREAM_IO(s)->print("%s: %s\r\n", key, value);
+  return dump_etag(s, boost::string_ref(bl_etag.c_str(), bl_etag.length()));
 }
 
 void dump_bucket_from_state(struct req_state *s)
 {
-  int expose_bucket = g_conf->rgw_expose_bucket;
-  if (expose_bucket) {
-    if (!s->bucket_name.empty()) {
-      string b;
-      if (!s->bucket_tenant.empty()) {
-        string g = s->bucket_tenant + "/" + s->bucket_name;
-        url_encode(g, b);
-      } else {
-        url_encode(s->bucket_name, b);
-      }
-      STREAM_IO(s)->print("Bucket: %s\r\n", b.c_str());
+  if (g_conf->rgw_expose_bucket && ! s->bucket_name.empty()) {
+    if (! s->bucket_tenant.empty()) {
+      dump_header(s, "Bucket",
+                  url_encode(s->bucket_tenant + "/" + s->bucket_name));
+    } else {
+      dump_header(s, "Bucket", url_encode(s->bucket_name));
```

Comment:
are you sure it's safe to pass a temporary `std::string` into `dump_header()` as a `const string_ref&` argument? it seems like the `string_ref` would end up with a pointer to freed memory